src/nunit.analyzers/ClassicModelAssertUsage/ClassicModelAssertUsageAnalyzer.cs

```diff
@@ -232,21 +233,58 @@ public sealed class ClassicModelAssertUsageAnalyzer : BaseAssertionAnalyzer
               { NameOfAssertIsNotInstanceOf, isNotInstanceOfDescriptor },
           }.ToImmutableDictionary();
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ClassicModelAssertUsageAnalyzer.name.Values.ToImmutableArray();
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ClassicModelAssertUsageAnalyzer.NameToDescriptor.Values.ToImmutableArray();
 
         protected override void AnalyzeAssertInvocation(SyntaxNodeAnalysisContext context,
             InvocationExpressionSyntax assertExpression, IMethodSymbol methodSymbol)
         {
-            if (ClassicModelAssertUsageAnalyzer.name.ContainsKey(methodSymbol.Name))
+            if (ClassicModelAssertUsageAnalyzer.NameToDescriptor.TryGetValue(methodSymbol.Name, out DiagnosticDescriptor? descriptor))
             {
+                string GetImplicitTypeConversion(ExpressionSyntax? expression)
+                {
+                    if (expression == null)
+                    {
+                        return string.Empty;
+                    }
+
+                    var typeInfo = context.SemanticModel.GetTypeInfo(expression, context.CancellationToken);
+
+                    var convertedType = typeInfo.ConvertedType;
+
+                    if (convertedType == typeInfo.Type)
+                    {
+                        return string.Empty;
+                    }
+
+                    if (convertedType.SpecialType == SpecialType.System_Object ||
```

Comment:
What is the logic here? Why object, IEnumerable and ICollection?

Reply:
object is the parameter type for Assert.That constraints and we don't need to cast to that.

ICollection is the parameter type for Assert.Contains.
IEnumerable is the fallback parameter type for Assert.Is(Not)Empty.

These are to prevent that `Assert.IsEmpty(myCollection)` becomes `Assert.That((IEnumerable)myCollection, Is.Empty)`.
and `Assert.Contains(myRecord, myCollection)` becomes `Assert.That((ICollection)myCollection, Does.Contain((object)myRecord)`

I didn't want to restrict it to only do the conversion logic for the comparable set of asserts: `Less`, `LessOrEqual`, `Greater`, `GreaterOrEqual` as I noticed that `IsTrue` and `IsFalse`  also need an explicit cast in the constraint model. Similar for `IsEmpty` and `IsNotEmpty` when the `string` overload is called.

I added more test cases for the 4 other asserts needing casts.

Reply:
I would suggest to get `compilation.ClassifyConversion(typeInfo.Type, convertedType)`, and checking for `IsUserDefined` property.
Would it work here?

Reply:
Thanks @Dreamescaper 
I tried `semanticModel.ClassifyConversion(expression, convertedType)` and it works! 
Allows me to remove these akward exclusions.